### PR TITLE
Add initial analytical memory estimator for DualPipeV training

### PR DESCRIPTION
**Pull request by MasterJH5574**
Initial version of a CPU-only tool that simulates the full DualPipeV 8-step pipeline schedule and estimates peak GPU memory per rank. Tracks all memory components: unsharded parameters, FSDP shards, optimizer states, per-chunk activations, autograd saved tensors, gradients, communication buffers, weight-grad-store deferred tensors, and non-PyTorch overhead (CUDA context, NCCL, torch.compile).

The activation model accounts for the MoE token expansion through EP all-to-all dispatch: each token selects k experts, yielding m*k total (token, expert) pairs at the scattered GEMM dimension, plus a 1.2x routing variance factor for worst-case layers. Transient tensors freed via deferred_free (sorted_tokens at stage3, moe_outs at stage5) are correctly excluded from persistent per-chunk memory. Storage sharing between autograd saved tensors and stage-boundary records is deducted to avoid double-counting.

Validated against real H100 measurements (pp=4, ep=8, 32 GPUs, Qwen3-30B-A3B): estimator predicts 66.6 GiB vs measured 62.3 GiB for pp_rank=0 (7% conservative margin) and 58.6 GiB vs 58.0 GiB for pp_rank=1 (1% error).

Includes an initial Claude Code skill (`/estimate-memory`) that translates natural-language queries into the correct CLI invocation, and static tests under tools/tests/.

This is the initial version. FP8 memory overhead, context parallelism (ring attention buffers), and DeepSeek-V2 shared experts are not yet modeled and will be improved in future iterations.

### CODE REVIEW

File: .claude/skills/estimate-memory/SKILL.md [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+The tool computes `dp_size = total_gpus / (pp_size * cp_size * ep_size)`. Verify this is an integer. If not, the config is invalid — tell the user.
+
+### Common hardware specs
+
+| Hardware | `--gpu-memory-gb` |
+|---|---|
+| H100 SXM | 80 |
+| H200 SXM | 141 |
+| A100 (80 GB) | 80 |
+| A100 (40 GB) | 40 |
```
Comment (haok1402) [inline .claude/skills/estimate-memory/SKILL.md]: @MasterJH5574 We don't support training on Ampere devices; probably the best to leave them out.

**Review by haok1402 — APPROVED** [review]

File: .claude/skills/estimate-memory/SKILL.md (hunk 1 above)
Comment (MasterJH5574) [reply .claude/skills/estimate-memory/SKILL.md]: Good catch.  Will update soon.

**Comment by MasterJH5574** [comment]
@haok1402 All comments addressed. Do you mind taking another look?

**Comment by haok1402** [comment]
@MasterJH5574 LGTM!
